src/mscorlib/shared/System/Memory.cs

```diff
@@ -294,7 +292,7 @@ public Span<T> Span
                 }
                 else if (_object != null)
                 {
-                    return new Span<T>((T[])_object, _index, _length & RemoveFlagsBitMask);
+                    return Span<T>.DangerousCreate(Unsafe.As<T[]>(_object), _index, _length & RemoveFlagsBitMask);
```

Comment:
I don't think we should do this. Previously this check provided a level of safety against a Memory being torn, where at least validation would be performed when getting out the span to ensure that the created span was still within the bounds of the array.  By removing those checks, races on reading/writing a memory in some shared field could result in getting out a span that points to unexpected places.

Reply:
Agree with @stephentoub . I do not think we should do this.

Reply:
How do we assess the risk of struct tearing in such cases given there are other places, particularly in Memory, where we don't guard against this? We don't check if the object field and index/length are in sync while slicing a memory, and even when getting the span property for objects other than T[].

Since you guys think that at the point of accessing the span, we should keep this check, I will revert this.

Reply:
The design point of the framework has always been that threading bugs can get you into invalid state, but they should not cause type safety or memory safety violations.